File: code/projects/08/clojure/src/Temperature.clj
;; TemperatureTable.cljis a module for Temperautres
;;
;; Usage: clojure -m TemperatureTable
;;
;; Completed by: Duncan Van Keulen
;; Date: 4/12/2020
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(require '[clojure.string :as str]) ;; include the Temperature module

;; Define a Temperature record with degree and scale attributes
(defrecord Temperature [myDegree myScale])

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; make-Temperature constructs a Temperature from a float and a character 
;;; Receive: degree : the degree of the temperature
;;; Recieve: scale : the scale that that degree is in
;;; Return: the Temperature (degree scale)
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn make-Temperature [degree scale]
  (->Temperature degree scale)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getDegree returns the degree of the recieved temperature object
;;; Receive: temp : a temperature
;;; PRE: temp MUST be a temperature that's initialized
;;; Return: the myDegree attribute of that temp object
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn getDegree [^Temperature temp] 
  (:myDegree temp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getScale returns the scale of the recieved temperature object
;;; Receive: temp : a temperature
;;; PRE: temp MUST be a temperature that's initialized
;;; Return: the myScale attribute of that temp object
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn getScale [^Temperature temp]
  (:myScale temp)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; isValidTemperature determines if a degree and scale would 
;;;     constitute a valid Temperature
;;; Receive: degree : the deree of the desired Temperature
;;; Recieve: scale : the scale of the desired Temperature
;;; Return: a boolean indicating whether or not the values provided
;;;     constitute a valid Temperature
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn isValidTemperature [degree scale] 
  (cond
    (= \F scale) (if (< degree -460.0) false true)
    (= \f scale) (if (< degree -460.0) false true)
    (= \C scale) (if (< degree -273.15) false true)
    (= \c scale) (if (< degree -273.15) false true)
    (= \K scale) (if (< degree 0.0) false true)
    (= \k scale) (if (< degree 0.0) false true)
    :else false
  )   
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; toFahrenheit converts a temperature to Fahrenheit from an arbitrary scale
;;; Receive: temp : the temperature to be converted
;;; Precondition: temp MUST be a Temperature and must be initialized
;;; Return: A new Temperature record with converted degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn toFahrenheit [^Temperature temp]
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) temp ;; no conversion needed
    (= \f (getScale temp)) temp ;; no conversion needed

    ;; From Celsius
    (= \C (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (getDegree temp)) 32.0) \F)
    (= \c (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (getDegree temp)) 32.0) \F)

    ;; From Kelvin
    (= \K (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (- (getDegree temp) 273.15)) 32.0) \F)
    (= \k (getScale temp)) (make-Temperature (+ (* (/ 9.0 5.0) (- (getDegree temp) 273.15)) 32.0) \F)

    :else false
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; toCelsius converts a temperature to Celsius from an arbitrary scale
;;; Receive: temp : the temperature to be converted
;;; Precondition: temp MUST be a Temperature and must be initialized
;;; Return: A new Temperature record with converted degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn toCelsius [^Temperature temp]
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) (make-Temperature (* (/ 5.0 9.0) (- (getDegree temp) 32.0)) \C)
    (= \f (getScale temp)) (make-Temperature (* (/ 5.0 9.0) (- (getDegree temp) 32.0)) \C)

    ;; From Celsius
    (= \C (getScale temp)) temp ;; no conversion needed
    (= \c (getScale temp)) temp ;; no conversion needed

    ;; From Kelvin
    (= \K (getScale temp)) (make-Temperature (- (getDegree temp) 273.15) \C)
    (= \k (getScale temp)) (make-Temperature (- (getDegree temp) 273.15) \C)

    :else false
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; toKelvin converts a temperature to Kelvin from an arbitrary scale
;;; Receive: temp : the temperature to be converted
;;; Precondition: temp MUST be a Temperature and must be initialized
;;; Return: A new Temperature record with converted degree and scale
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn toKelvin [^Temperature temp] 
  (cond
    ;; From Fahrenheit
    (= \F (getScale temp)) (make-Temperature (+ (* (- (getDegree temp) 32.0) (/ 5.0 9.0)) 273.15) \K)
    (= \f (getScale temp)) (make-Temperature (+ (* (- (getDegree temp) 32.0) (/ 5.0 9.0)) 273.15) \K)

    ;; From Celsius
    (= \C (getScale temp)) (make-Temperature (+ (getDegree temp) 273.15) \K)
    (= \c (getScale temp)) (make-Temperature (+ (getDegree temp) 273.15) \K)

    ;; From Kelvin
    (= \K (getScale temp)) temp ;; no conversion needed
    (= \k (getScale temp)) temp ;; no conversion needed

    :else false
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; enterTemperature allows a Temperature record's values to be 
;;;     entered by the user
;;; Recieve: degree and scale to be read in as input
;;; Precondition: the input degree and scale must consitute a
;;;     valid Temperature object
;;; Return: a new Temperature record with the user entered values
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn enterTemperature []
  (let 
    [
      input (read-line) ;; get the input from the user
      splicedInput (str/split input #" ")
      inputDegree (Float/parseFloat (first splicedInput))
      inputScale (get (second splicedInput) 0)
    ]
    (if (isValidTemperature inputDegree inputScale)
      (->Temperature inputDegree inputScale)
      (do (println "Couldn't enter temperature") (flush))
    )
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; dispalyTemperature prints a Temperature record to the screen
;;; Receive: temp : a Temperature
;;; Precondition: temp MUST be an initialized Temperature record
;;; Postcondition: the Temperature will be conveniently formatted
;;;     and printed to the screen
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn displayTemperature [^Temperature temp]
  (print (format "%.5f"(getDegree temp)) " " (getScale temp))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; raiseTemperature raises a temperature's degree attribute by
;;;     the input degrees
;;; Receive: temp : a Temperature record
;;; Recieve: degree : a float - the desired degrees to raise by
;;; Precondition: temp MUST be an initialized Temperature record
;;; Precondition: raising by degrees must consitute a valid Temperature
;;; Return: a new temperature with updated degrees
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn raiseTemperature [^Temperature temp degree]
  (if (isValidTemperature (+ (getDegree temp) degree) (getScale temp)) 
    ;; True
    (make-Temperature (+ (getDegree temp) degree) (getScale temp))
    ;; False
    (do (print "Could not raise temperature") (flush))
  )  
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; lowerTemperature lowers a temperature's degree attribute by
;;;     the input degrees
;;; Receive: temp : a Temperature record
;;; Recieve: degree : a float - the desired degrees to lower by
;;; Precondition: temp MUST be an initialized Temperature record
;;; Precondition: raising by degrees must consitute a valid Temperature
;;; Return: a new temperature with updated degrees
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn lowerTemperature [^Temperature temp degree]
  (if (isValidTemperature (- (getDegree temp) degree) (getScale temp)) 
    ;; True
    (make-Temperature (- (getDegree temp) degree) (getScale temp))
    ;; False
    (do (print "Could not lower temperature") (flush))
  )  
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; equals allows checking if two Temperatures are equal regardless of scale
;;; Receive: temp1 : Temperature - the lhs of the = operation
;;; Recieve: temp2 : Temprature - the rhs of the = operation
;;; Precondition: temp1 & temp2 MUST be initialized Temperature records
;;; Return: a boolean indicating whether or not the temperatures are equal
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn equals [^Temperature temp1 ^Temperature temp2]
  (if (= (getDegree (toFahrenheit temp1)) (getDegree (toFahrenheit temp2)))
    ;; True-Section
      true ;; return true
    ;; False-Section
      false ;; return false
  )
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; lessThan allows checking if one Temperature is less than another regardless of scale
;;; Receive: temp1 : Temperature - the lhs of the < operation
;;; Recieve: temp2 : Temprature - the rhs of the < operation
;;; Precondition: temp1 & temp2 MUST be initialized Temperature records
;;; Return: a boolean indicating if temp1 < temp2
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn lessThan [^Temperature temp1 ^Temperature temp2]
  (if (< (getDegree (toFahrenheit temp1)) (getDegree (toFahrenheit temp2)))
    ;; True-Section
      true ;; return true
    ;; False-Section
      false ;; return false
  )
)

File: code/projects/08/java/Temperature.java
/* Tempearture.java provides the Temperature class
* Student: Duncan Van Keulen
* Date: 4/9/2020
*************************************************/

public class Temperature {
    private Double myDegrees;
    private char myScale;

    /*
    Temperature constructor
    In: double degrees: the value of the degrees to be used
    In: char scale: the temperature scale to use
    Precondition: Scale must be one of C, K, F/c, k, f
    Precondition: Degrees must not be below absolute zero in the respective scale
    Postcondition: Temperature object will be made and instance variables
                    will be initialized
    */
    public Temperature(final double degrees, final char scale) {
        if(isValidTemperature(degrees, scale)) {
            myScale = scale;
            myDegrees = degrees;
        }
    }

    /*
    Empty temperature constructor for use with filling values in later
    */
    public Temperature() {
        myScale = 'C';
        myDegrees = 0.0;
    };

    /* 
    Checker to make sure a temperature is going to be valid
    In: double degree: the degree of the perspective Temperature
    In: char scale: the temperature scale of the perspective Temperature
    Return: Boolean indicating whether or not the temperature is valid. 
    */
    public boolean isValidTemperature(double degree, char scale) {
        // Make sure the temperature scale is one of these
        switch(scale) {
            case 'F':
            case 'f':
                if(degree >= -460.0) {
                    return true;
                }
            case 'K':
            case 'k':
                if (degree >= 0.0) {
                    return true;
                }
            case 'C':
            case 'c':
                if (degree >= -273.15) {
                    return true;
                }
            default:
                return false;
        }
    }

    // Accessor for myScale
    public char getScale() {
        return myScale;
    }

    // Accessor of myDegrees
    public double getDegrees() {
        return myDegrees;
    }

    /* 
    Upwards adjustment for degrees
    In: Temperature temp: the temperature object of which to raise the degrees by
    In: double degrees: the amount of degrees to raise the temperature by in it's scale
    Precondition: temp must be a valid temperature.
    Returns: A new temperature object with the myDegree incremented by degrees
    */
    public Temperature raise(Temperature temp, double degrees) {
        double newDeg = temp.getDegrees();
        newDeg += degrees;
        if(isValidTemperature(newDeg, temp.getScale())){
            return new Temperature(newDeg, temp.getScale());
        }
        else {
            System.out.println("Couldn't raise temperature");
            return temp;
        }
    }

    /* 
    Downwards adjustment for degrees
    In: Temperature temp: the temperature object of which to lower the degrees by
    In: double degrees: the amount of degrees to lower the temperature by in it's scale
    Precondition: temp must be a valid temperature.
    Returns: A new temperature object with the myDegree decremented by degrees
    */
    public Temperature lower(Temperature temp, double degrees) {

        double newDeg = temp.getDegrees();
        newDeg -= degrees;

        if(isValidTemperature(newDeg, temp.getScale())){
            return new Temperature(newDeg, temp.getScale());
        }
        else {
            System.out.println("Couldn't raise temperature");
            return temp;
        }
    }

    /* 
    Fahrenheit conversion
    Converty any temperature scale into Fahrenheit. 
    Returns: A new Temperature object with the myScale
            changed to the correct scale and myDegree
            updated to the correct value for that scale
    */
    public Temperature toFahrenheit() {
        switch(myScale) {
            case 'F':
            case 'f':
                return this;
            case 'C':
            case 'c':
                return new Temperature( ((myDegrees * (9.0/5.0)) + 32.0), 'F');
            case 'K':
            case 'k':
                return new Temperature( (((myDegrees - 273.15) * (9.0/5.0)) + 32.0), 'F' );
            default:
                // Invalid temperature scale
                System.out.println("Could not convert to F");
                return this;
        }
    }

    /* 
    Celsius conversion
    Converty any temperature scale into Celsius. 
    Returns: A new Temperature object with the myScale
            changed to the correct scale and myDegree
            updated to the correct value for that scale
    */
    public Temperature toCelsius() {
        switch(myScale) {
            case 'F':
            case 'f':
                return new Temperature( (myDegrees - 32.0) * (5.0/9.0), 'C');
            case 'C':
            case 'c':
                return this;
            case 'K':
            case 'k':
                return new Temperature((myDegrees - 273.15), 'C');
            default:
                // Invalid temperature scale
                System.out.println("Could not convert to C");
                return this;
        }
    }

    /* 
    Kelvin conversion
    Converty any temperature scale into Kelvin. 
    Returns: A new Temperature object with the myScale
            changed to the correct scale and myDegree
            updated to the correct value for that scale
    */
    public Temperature toKelvin() {
        switch(myScale) {
            case 'F':
            case 'f':
                return new Temperature( (((myDegrees - 32.0)) * (5.0/9.0) + 273.15), 'K');
            case 'C':
            case 'c':
                return new Temperature( (myDegrees + 273.15), 'K');
            case 'K':
            case 'k':
                return this;
            default:
                // Invalid temperature scale
                System.out.println("Could not convert to K");
                return this;
        }
    }

    /*
    Ability to set the values of myDegree and myScale with a string input
    In: String input: the string containing the desired temperature and scale
    Precondition: input must be in the format <valid numbers> <valid scale>
                and must be valid for a Temperature object
    Postcondition: The values of myDegree and myScale will be updated with the input values
    */
    public void enterTemperature(String input) {
        // Strip away nasty white space
        char[] inputCharArray = input.toCharArray();

        // Set the scale to the last character entered
        char scale = inputCharArray[input.length() - 1];

        // Set the degrees to everthing before that (in hopes that it's a valid double)
        double degrees = Double.parseDouble( input.substring(0, input.length() - 1) );

        // Make sure the parsed values would make a valid temperature and set 
        //      myDegree and myScale
        if(isValidTemperature(degrees, scale)) {
            myDegrees = degrees;
            myScale = scale;
        }
        else {
            System.out.println("Invalid temperature");
        }
    }

    /*
    Convert a temperature object to a string to be displayed (in a table)
    Returns: A string containing the value of myDegrees rounded to 2 decimal places
            followed by the letter corresponding to the scale
    */
    public String toString() {
        return String.format(("%.2f " + String.valueOf(myScale)), myDegrees);
    }

    /*
    For testing if two Temperature objects have equal temperature (regardless of scale)
    In: Temperature temp: The Temperature object to compare to this.
    Precondition: Both temp and this must be valid temperature objects.
    Returns: a boolean indicating whether or not the two temperatures are equal.
    */
    public boolean equals(Temperature temp) {
        switch(myScale) {
            case 'F':
            case 'f':
                if(temp.toFahrenheit().getDegrees() == myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            case 'K':
            case 'k':
                if(temp.toKelvin().getDegrees() == myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            case 'C':
            case 'c':
                if(temp.toCelsius().getDegrees() == myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            default:
                return false;
        }
    }

    /*
    For testing if this Temperature object's degree is less than another
    In: Temperature temp: The Temperature object to compare to this.
    Precondition: Both temp and this must be valid temperature objects.
    Returns: a boolean indicating whether or not temp's degree is less than this's degree.
    */
    public boolean lessThan(Temperature temp) {
        switch(myScale) {
            case 'F':
            case 'f':
                if(temp.toFahrenheit().getDegrees() > myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            case 'K':
            case 'k':
                if(temp.toKelvin().getDegrees() > myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            case 'C':
            case 'c':
                if(temp.toCelsius().getDegrees() > myDegrees) {
                    return true;
                }
                else {
                    return false;
                }
            default:
                return false;
        }
    }
}
